From bbb3a242d62ddeaa76dcbff3ce0ac1e387b43254 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:35:23 -0400
Subject: [block] blk-mq: bidi support

Message-id: <1402670238-13916-36-git-send-email-snitzer@redhat.com>
Patchwork-id: 83793
O-Subject: [RHEL7.1 PATCH 035/150] blk-mq: bidi support
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 91b63639c7d59dbf86da5311faff43691dcae832
Author: Christoph Hellwig <hch@lst.de>
Date:   Wed Apr 16 09:44:53 2014 +0200

    blk-mq: bidi support

    Add two unlinkely branches to make sure the resid is initialized correctly
    for bidi request pairs, and the second request gets properly freed.

    Signed-off-by: Christoph Hellwig <hch@lst.de>
    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 86d66e0..963a821 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -298,10 +298,13 @@ inline void __blk_mq_end_io(struct request *rq, int error)
 {
  blk_account_io_done(rq);
 
- if (rq->end_io)
+ if (rq->end_io) {
   rq->end_io(rq, error);
- else
+ } else {
+  if (unlikely(blk_bidi_rq(rq)))
+   blk_mq_free_request(rq->next_rq);
   blk_mq_free_request(rq);
+ }
 }
 EXPORT_SYMBOL(__blk_mq_end_io);
 
@@ -366,6 +369,8 @@ static void blk_mq_start_request(struct request *rq, bool last)
  trace_block_rq_issue(q, rq);
 
  rq->resid_len = blk_rq_bytes(rq);
+ if (unlikely(blk_bidi_rq(rq)))
+  rq->next_rq->resid_len = blk_rq_bytes(rq->next_rq);
 
  /*
   * Just mark start time and set the started bit. Due to memory
-- 
1.7.1